docs/development/pigs/pig-007.rst

```diff
@@ -0,0 +1,409 @@
+prototype so that the following easily works:
+
+.. code-block::
+
+    from gammapy.cube import SourceModels
+
+    model = SourceModels.read("model.xml")
+    print(model)
+
+    model.write("model.xml")
+
+
+In addition we propose to implement an additional YAML serialization format, which
+results in human-readable model configuration files. Once available the YAML format
+should be the preferred serialization:
+
+.. code-block::
+
+    model = SourceModels.read("model.yaml")
+    print(model)
+
+    model.write("model.yaml")
+
+The two file formats should be handled automatically in `SourceModel.read()` and
+`SourceModel.write()`.
+
+
+Improve spatial models
+----------------------
+
+Implement sky coordinate handling
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+Implement support for coordinate systems. Add an interface for `SkyCoord`, by
+introducing a `SpatialModel.skycood` attribute.
+
+.. code-block::
+
+    from astropy.coordinates import SkyCoord
+    from gammapy.image.models import PointSource
+
+    # instanciation using a SkyCoord object
+    point_source = PointSource(skycood=SkyCoord(0, 0, frame="galactic", unit="deg"))
+    # or
+    point_source = PointSource(lon="0 deg", lat="0 deg", frame="galactic")
+
+    skycoord = point_source.skycoord
+    assert isinstance(skycoord, SkyCoord)
+
+    # evaluation of models using sky coordinates
+    coords = SkyCoord([0, 1], [0.5, 0.5], frame="galactic", unit="deg")
+    values = point_source(coords)
+
+    # maybe override evaluation for efficient evaluation for fitting
+    values = point_source(lon, lat)
+
+
+Implement default parameters
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+For spectral models Gammapy has defined default parameters, which simplifes the
+interactive usage of the models. From this experience we propose to introduce
+default parameters for `SpatialModel` as well as `SourceModel`.
+
+
+.. code-block::
+
+    from gammapy.image.models import Shell
+    from gammapy.cube.models import SourceModels
+
+    shell = Shell()
+    print(shell)
+
+
+Implement evaluation region specifications
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+For efficient evaluation of model components, we propose that `SpatialModel`
+classes declare a circular region `.evaluation_radius`, they are evaluated on.
+
+.. code-block::
+
+    from gammapy.image.models import Shell
+
+    shell = Shell()
+    print(shell.evaluation_radius)
+
+The `.evaluation_radius` should be a property of the model, that is computed
+depending of the model parameter, that determines the spatial extension of the
+model. The handling of `PointSource` and additional evaluation margins, which depend
+on the pixel size, is done in the step, where the model component is evaluated.
+
+
+Expose model parameters as attributes
+-------------------------------------
+To simplify the interactive handling of model instances, model parameters should
+be exposed as attribues on the model, so that the following works:
```

Comment:
Should we just require model and parameter names that are valid Python identifiers?
Or are other names like "Axel's discovery!" also allowed?
Suggest to be clear on this point.
(I think it's OK to require valid Python identifier, even if supporting any string would be possible.)
If the requirement is made, an error should be thrown as soon as something invalid arrives.
In the PIG it's just one sentence, but I think it's good to clarify.

Concerning exposing parameters (or presumably model parts) as attributes on models.
I can see the appeal, but it also leads to problems because it mixes into the namespace of methods or other attributes on those classes.
So if we expose this, we should be clear on what happens if there is a name conflict. E.g. pandas does this on dataframe, and as soon as there is a name conflict, that column is no longer accessible as an attribute.
We could also generate an error if such a name conflict occurs (limiting the set of allowed parameter names a bit, e.g. forbidding a name like "energy_flux" or "spectral_index" on a spectral model because it's already a method name.
A detail, but a controversial detail that needs to be discussed / decided, so that whoever works on this next year can just go ahead and implement this, and PR discussions will revolve around implementation & tests & docs, and no longer "what do we want?".